Support IAM Roles for Tasks when running in AWS ECS

Spring Cloud has support for using the [IAM instance profile](http://cloud.spring.io/spring-cloud-aws/spring-cloud-aws.html#_instance_profile_configuration) for credential credential retrieval when running in AWS. However, when attempting to use AWS ECS [IAM roles for tasks](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) the application will still use the host instance profile.

As far as I understand the difference between an instance profile and a task role should be transparent for AWS SDK for Java post version 1.11.16 so I assuming there are some explicit controls in Spring Cloud that prevent IAM roles for tasks from working.

See below the application has a ECS task role `
inventory-canary-InventoryTaskRole` but is using the container host's instance profile role of `inventory-canary-EC2Role`, and the AWS JDK version is post 1.11.16.

```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.cloud.aws.core.env.stack.config.StackResourceRegistryFactoryBean]: Factory method 'stackResourceRegistryFactoryBean' threw exception; nested exception is com.amazonaws.AmazonServiceException: User: arn:aws:sts::XXXXXXXXXXXX:assumed-role/inventory-canary-EC2Role-10RJ4L9FL173K/i-0a9f98379c2f38e45 is not authorized to perform: cloudformation:DescribeStackResources (Service: AmazonCloudFormation; Status Code: 403; Error Code: AccessDenied; Request ID: 11f41133-e80f-11e6-a093-4737c2e1094e)
at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189) ~[spring-beans-4.3.5.RELEASE.jar!/:4.3.5.RELEASE]
at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588) ~[spring-beans-4.3.5.RELEASE.jar!/:4.3.5.RELEASE]
... 106 common frames omitted
Caused by: com.amazonaws.AmazonServiceException: User: arn:aws:sts::614382051240:assumed-role/inventory-canary-EC2Role-10RJ4L9FL173K/i-0a9f98379c2f38e45 is not authorized to perform: cloudformation:DescribeStackResources (Service: AmazonCloudFormation; Status Code: 403; Error Code: AccessDenied; Request ID: 11f41133-e80f-11e6-a093-4737c2e1094e)
at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:1378) ~[aws-java-sdk-core-1.11.18.jar!/:na]
at com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:924) ~[aws-java-sdk-core-1.11.18.jar!/:na]
at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:702) ~[aws-java-sdk-core-1.11.18.jar!/:na]
at com.amazonaws.http.AmazonHttpClient.doExecute(AmazonHttpClient.java:454) ~[aws-java-sdk-core-1.11.18.jar!/:na]
at com.amazonaws.http.AmazonHttpClient.executeWithTimer(AmazonHttpClient.java:416) ~[aws-java-sdk-core-1.11.18.jar!/:na]
at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:365) ~[aws-java-sdk-core-1.11.18.jar!/:na]
at com.amazonaws.services.cloudformation.AmazonCloudFormationClient.doInvoke(AmazonCloudFormationClient.java:1633) ~[aws-java-sdk-cloudformation-1.11.18.jar!/:na]
at com.amazonaws.services.cloudformation.AmazonCloudFormationClient.invoke(AmazonCloudFormationClient.java:1603) ~[aws-java-sdk-cloudformation-1.11.18.jar!/:na]
at com.amazonaws.services.cloudformation.AmazonCloudFormationClient.describeStackResources(AmazonCloudFormationClient.java:892) ~[aws-java-sdk-cloudformation-1.11.18.jar!/:na]
at org.springframework.cloud.aws.core.env.stack.config.AutoDetectingStackNameProvider.autoDetectStackName(AutoDetectingStackNameProvider.java:76) ~[spring-cloud-aws-core-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
at org.springframework.cloud.aws.core.env.stack.config.AutoDetectingStackNameProvider.afterPropertiesSet(AutoDetectingStackNameProvider.java:62) ~[spring-cloud-aws-core-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
```

--- a/spring-cloud-aws-autoconfigure/src/test/java/org/springframework/cloud/aws/autoconfigure/context/ContextCredentialsAutoConfigurationTest.java
+++ b/spring-cloud-aws-autoconfigure/src/test/java/org/springframework/cloud/aws/autoconfigure/context/ContextCredentialsAutoConfigurationTest.java
@@ -19,7 +19,7 @@ package org.springframework.cloud.aws.autoconfigure.context;
 import com.amazonaws.auth.AWSCredentialsProvider;
 import com.amazonaws.auth.AWSCredentialsProviderChain;
 import com.amazonaws.auth.AWSStaticCredentialsProvider;
-import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper;
 import com.amazonaws.auth.profile.ProfileCredentialsProvider;
 import org.apache.http.client.CredentialsProvider;
 import org.junit.After;
@@ -67,7 +67,7 @@ public class ContextCredentialsAutoConfigurationTest {
 		@SuppressWarnings("unchecked") List<CredentialsProvider> credentialsProviders =
 				(List<CredentialsProvider>) ReflectionTestUtils.getField(awsCredentialsProvider, "credentialsProviders");
 		assertEquals(2, credentialsProviders.size());
-		assertTrue(InstanceProfileCredentialsProvider.class.isInstance(credentialsProviders.get(0)));
+		assertTrue(EC2ContainerCredentialsProviderWrapper.class.isInstance(credentialsProviders.get(0)));
 		assertTrue(ProfileCredentialsProvider.class.isInstance(credentialsProviders.get(1)));
 	}
 
@@ -106,7 +106,7 @@ public class ContextCredentialsAutoConfigurationTest {
 		@SuppressWarnings("unchecked") List<CredentialsProvider> credentialsProviders =
 				(List<CredentialsProvider>) ReflectionTestUtils.getField(awsCredentialsProvider, "credentialsProviders");
 		assertEquals(2, credentialsProviders.size());
-		assertTrue(InstanceProfileCredentialsProvider.class.isInstance(credentialsProviders.get(0)));
+		assertTrue(EC2ContainerCredentialsProviderWrapper.class.isInstance(credentialsProviders.get(0)));
 		assertTrue(ProfileCredentialsProvider.class.isInstance(credentialsProviders.get(1)));
 	}
 
@@ -123,7 +123,7 @@ public class ContextCredentialsAutoConfigurationTest {
 		@SuppressWarnings("unchecked") List<CredentialsProvider> credentialsProviders =
 				(List<CredentialsProvider>) ReflectionTestUtils.getField(awsCredentialsProvider, "credentialsProviders");
 		assertEquals(2, credentialsProviders.size());
-		assertTrue(InstanceProfileCredentialsProvider.class.isInstance(credentialsProviders.get(0)));
+		assertTrue(EC2ContainerCredentialsProviderWrapper.class.isInstance(credentialsProviders.get(0)));
 		assertTrue(ProfileCredentialsProvider.class.isInstance(credentialsProviders.get(1)));
 
 		assertEquals("test", ReflectionTestUtils.getField(credentialsProviders.get(1), "profileName"));
@@ -143,7 +143,7 @@ public class ContextCredentialsAutoConfigurationTest {
 		@SuppressWarnings("unchecked") List<CredentialsProvider> credentialsProviders =
 				(List<CredentialsProvider>) ReflectionTestUtils.getField(awsCredentialsProvider, "credentialsProviders");
 		assertEquals(2, credentialsProviders.size());
-		assertTrue(InstanceProfileCredentialsProvider.class.isInstance(credentialsProviders.get(0)));
+		assertTrue(EC2ContainerCredentialsProviderWrapper.class.isInstance(credentialsProviders.get(0)));
 		assertTrue(ProfileCredentialsProvider.class.isInstance(credentialsProviders.get(1)));
 
 		ProfileCredentialsProvider provider = (ProfileCredentialsProvider) credentialsProviders.get(1);

--- a/spring-cloud-aws-context/src/main/java/org/springframework/cloud/aws/context/config/support/ContextConfigurationUtils.java
+++ b/spring-cloud-aws-context/src/main/java/org/springframework/cloud/aws/context/config/support/ContextConfigurationUtils.java
@@ -18,7 +18,7 @@ package org.springframework.cloud.aws.context.config.support;
 
 import com.amazonaws.auth.AWSStaticCredentialsProvider;
 import com.amazonaws.auth.BasicAWSCredentials;
-import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper;
 import com.amazonaws.auth.profile.ProfileCredentialsProvider;
 import org.springframework.beans.factory.config.BeanDefinition;
 import org.springframework.beans.factory.config.BeanDefinitionHolder;
@@ -85,7 +85,7 @@ public final class ContextConfigurationUtils {
 		}
 
 		if (instanceProfile) {
-			awsCredentialsProviders.add(BeanDefinitionBuilder.rootBeanDefinition(InstanceProfileCredentialsProvider.class).getBeanDefinition());
+			awsCredentialsProviders.add(BeanDefinitionBuilder.rootBeanDefinition(EC2ContainerCredentialsProviderWrapper.class).getBeanDefinition());
 		}
 
 		if (StringUtils.hasText(profileName)) {

--- a/spring-cloud-aws-context/src/test/java/org/springframework/cloud/aws/context/config/annotation/ContextCredentialsConfigurationRegistrarTest.java
+++ b/spring-cloud-aws-context/src/test/java/org/springframework/cloud/aws/context/config/annotation/ContextCredentialsConfigurationRegistrarTest.java
@@ -19,7 +19,7 @@ package org.springframework.cloud.aws.context.config.annotation;
 import com.amazonaws.auth.AWSCredentialsProvider;
 import com.amazonaws.auth.AWSStaticCredentialsProvider;
 import com.amazonaws.auth.DefaultAWSCredentialsProviderChain;
-import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper;
 import com.amazonaws.auth.profile.ProfileCredentialsProvider;
 import org.apache.http.client.CredentialsProvider;
 import org.junit.After;
@@ -156,7 +156,7 @@ public class ContextCredentialsConfigurationRegistrarTest {
 				(List<CredentialsProvider>) ReflectionTestUtils.getField(awsCredentialsProvider, "credentialsProviders");
 		assertEquals(2, credentialsProviders.size());
 		assertTrue(AWSStaticCredentialsProvider.class.isInstance(credentialsProviders.get(0)));
-		assertTrue(InstanceProfileCredentialsProvider.class.isInstance(credentialsProviders.get(1)));
+		assertTrue(EC2ContainerCredentialsProviderWrapper.class.isInstance(credentialsProviders.get(1)));
 	}
 
 	@Test
@@ -173,7 +173,7 @@ public class ContextCredentialsConfigurationRegistrarTest {
 		@SuppressWarnings("unchecked") List<CredentialsProvider> credentialsProviders =
 				(List<CredentialsProvider>) ReflectionTestUtils.getField(awsCredentialsProvider, "credentialsProviders");
 		assertEquals(1, credentialsProviders.size());
-		assertTrue(InstanceProfileCredentialsProvider.class.isInstance(credentialsProviders.get(0)));
+		assertTrue(EC2ContainerCredentialsProviderWrapper.class.isInstance(credentialsProviders.get(0)));
 	}
 
 	@Test
@@ -244,7 +244,7 @@ public class ContextCredentialsConfigurationRegistrarTest {
 				(List<CredentialsProvider>) ReflectionTestUtils.getField(awsCredentialsProvider, "credentialsProviders");
 		assertEquals(3, credentialsProviders.size());
 		assertTrue(AWSStaticCredentialsProvider.class.isInstance(credentialsProviders.get(0)));
-		assertTrue(InstanceProfileCredentialsProvider.class.isInstance(credentialsProviders.get(1)));
+		assertTrue(EC2ContainerCredentialsProviderWrapper.class.isInstance(credentialsProviders.get(1)));
 		assertTrue(ProfileCredentialsProvider.class.isInstance(credentialsProviders.get(2)));
 	}
 

